src/core/withExperiment.js

```diff
@@ -0,0 +1,76 @@
+/* @flow */
+import * as React from 'react';
+import invariant from 'invariant';
+import cookie from 'react-cookie';
+
+export type ExposedWithExperimentProps = {|
+  variant: string,
+|};
+
+type Props = {|
+  cookieConfig?: Object,
+  id: string,
+  variantA: string,
+  variantB: string,
+|};
+
+type InternalProps = {|
+  ...Props,
+  WrappedComponent: Function,
+  _cookie: typeof cookie,
+  randomizer: () => number,
+|};
+
+const defaultCookieConfig = { path: '/' };
+
+export const withExperiment = ({
+  cookieConfig = defaultCookieConfig,
+  id: defaultId,
+  variantA: defaultVariantA,
+  variantB: defaultVariantB,
+}: Props) => (WrappedComponent: Function) => {
+  class WithExperiment extends React.Component<InternalProps> {
+    experimentCookie: string | void;
+
+    static defaultProps = {
+      _cookie: cookie,
+      id: defaultId,
+      randomizer: Math.random,
+      variantA: defaultVariantA,
+      variantB: defaultVariantB,
+    };
+
+    constructor(props: InternalProps) {
+      super(props);
+
+      const { _cookie, id, randomizer, variantA, variantB } = this.props;
+
+      invariant(id, 'id is required');
+      invariant(variantA, 'variantA is required');
+      invariant(variantB, 'variantB is required');
```

Comment:
I think, if you move those `invariant` calls above the class, then we'll be able to validate the HOC call and we would not need those calls here (if that makes sense).

I'd put the `invariant` above the class, on the `default*` variables.

Reply:
oh?! ok sure I'll take a look..